Exchange Feature Info via Version Manager

stefaus opened the pull request:
The context of this PR is to add a generic way to share information like a Skype name outside of a session. Additionally add a platform info like eclipse or intellij to allow client checks.


### Current behavior
After xmpp connect a simple available presence update is send to the server.
When a available presence update of other contacts is retrieved, a service discovery via `DiscoveryService` is done. This returns a list of strings which is checked for 'saros' and 'saros.server'.
A version check is done via `VersionManager` at session start.

### Proposal
Remove the discovery service logic and use the `VersionManager` like [@srossbach already proposed](https://github.com/saros-project/saros/pull/766#issuecomment-558698991) to handle features. This allows to exchange key-value lists, include information like a Skype username and remove extra Skype info messages.

The `VersionManager` uses a saros specific message and a different client would just not response and this would work. But this behavior like the current discovery logic leads to a message flood to all available contacts and should be avoided.
[XEP-115](https://xmpp.org/extensions/xep-0115.html) proposes a way to exchange version information already via presence updates. This allows to get info about the client (saros), platform (like eclipse) and version string (like 15.0.0.dev) at time of contact availability. This limits the need to send version exchange requests to saros clients.

A thing to discuss is if the current version compatibility logic is needed.
The logic to check backwards compatibility in server→client and client→server direction may be working but is currently never used.
As @tobous mentioned a different qualifier should already make clients incompatible e.g. to avoid problems using different builds.
So a proposal would be to check if the complete version string matches and later if needed add a logic that allows compatibility within the same patch level or some other simple logic.


@srossbach @tobous @m273d15 any input on this?

saros-infrastructure commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- core/src/saros/versioning/ClientInfo.java  3
         

Complexity decreasing per file
==============================
+ eclipse/src/saros/communication/SkypeManager.java  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5072803&bid=16689379)

m273d15 commented:
> Remove the discovery service logic and use the VersionManager like @srossbach already proposed to handle features. This allows to exchange key-value lists, include information like a Skype username and remove extra Skype info messages.

I like the idea, but we should think about renaming and restructuring the code. A VersionExchangeExtension/VersionManager that is used to share the skype name (or other information that is shared in the future) might confuse developers. I think the perspective (and code structure) should be "It is a general feature for sharing client-specific information that is also used to share the version" instead of "It is a feature for sharing the version information that is also used to share general information".

> So a proposal would be to check if the complete version string matches and later if needed add a logic that allows compatibility within the same patch level or some other simple logic.

I think this solution would be sufficient for our current needs :+1:

stefaus commented:
> we should think about renaming and restructuring the code
Yes definitely! First thought (see todo remark at `VersionManager`) was to name it feature manager, but maybe better something in ClientInfoManager direction.

srossbach commented:
I do not like the approach of encoding / decoding the client capabilities (platform, version) into the node element of the presence extension. This is not the purpose of XEP-115.

The purpose of this extension is to generate a node and ver pair so you can lookup the supported features in your cache (did you even configure it ? see https://github.com/igniterealtime/Smack/blob/master/smack-extensions/src/main/java/org/jivesoftware/smackx/caps/EntityCapsManager.java#L265 but I do not know if Smack sets a default one) and so reduce packet sending and replies.

stefaus commented:
Yeah, I should have written that XEP-115 allows first of all to get a client name (in node element, which should be a URI) and a hash over service discovery values.
So according to the XEP draft this should just be something like "https://saros-project.org".
The hacky thing I done here is to add platform and version separated by whitespace.
If changed to a "/" we could use "https://saros-project.org/eclipse/15.0.0.dev" as URI and this would be more or less fine.

Using Caps Management just makes sense if service discovery is used. As as I understand your view this is the thing you wanted to avoid.
So my thought for including the version was to avoid the disco request. But as the version message is easy and should be stable just knowing that a client is a saros client is good enough.

---
So how about just "https://saros-project.org" as node, remove service discovery stuff and put platform and version into the version manager message?

srossbach commented:
I guess that it is the best approach. Using a simple node name to indicate that this is a Saros client (either a server or a normal one). The remaining data should be exchanged through IQ packets (or whatever the underlying network layer is), i.e there is central manager that broadcasts or requests several configurations (e.g version, skype name, whatever). We basically just have to decide if we push our data to other clients or if other clients will always pull us.

E.g you call VersionManager.requestVersion(contact) which will call some implementation of XYZFeatureBlaBla.getVersion(contact) which can then decide based on the node name if it even starts sending a packet to the destination or not.

stefaus commented:
I updated the code.
- as talked about in the Stand-Ups the Version Compatibility stuff is simplified to just check if the Version matches
  - to include a check for qualifier and platform is part of a later PR
- changed the Request/Get version handling to push the information to all contacts that are/become available and have saros support; This way a info like skypename is already available when the contact context menu is opened
- moved the skypename and server feature exchange to the VersionManager
- changed the check for saros availability from DiscoveryService to exchange via Caps Presence Extention
- removed stuff like DiscoveryService and ResourceFeature
- as compatibility with older versions is bricked here anyway, added NamedMapConverter to xstream for a shorter and better to read Map conversion
- Not included now for readability: would rename `VersionManager` to `InfoManager` and the package `saros.versioning` to `saros.communication.info`

stefaus commented on core/src/saros/communication/extensions/VersionExchangeExtension.java:
```diff
@@ -1,54 +0,0 @@
-package saros.communication.extensions;
```
yes this whole pr breaks compatibility, but @tobous, @m273d15 and i are fine with that

older clients will newer think about sending this packet because, as the `saros` feature is removed from service discovery, they will not see saros support.

stefaus commented on core/src/saros/misc/xstream/XStreamExtensionProvider.java:
```diff
@@ -99,6 +100,27 @@ public XStreamExtensionProvider(String namespace, String elementName, Class<?>..
     xstream.registerConverter(new UrlEncodingStringConverter());
+
+    /*
+     * set map conversion format to
```
not necessary, but because this pr breaks compatibility anyway a chance to increase readability of our xmpp messages and make them shorter

so already changing this breaks the whole compatibility because old `VersionExchangeExtension` couldn't be parsed

stefaus commented on core/src/saros/net/xmpp/XMPPConnectionService.java:
```diff
@@ -34,6 +38,8 @@
  */
 @Component(module = "net")
 public class XMPPConnectionService {
+  public static final String XMPP_CLIENT_IDENTIFIER = "https://saros-project.org";
```
I don't think so, if another protocol would use a different naming we wouldn't use it.
The resource part is also xmpp specific and should move to `saros.net.xmpp`

stefaus commented on core/src/saros/versioning/VersionManager.java:
```diff
@@ -1,330 +1,184 @@
+    else localInfo.put(key, value);
+
+    if (remoteInfo.isEmpty()) return;
+    dispatchThread.executeAsDispatch(
```
I just looked till
https://github.com/saros-project/saros/blob/f13bdc3ea5c5d295428c01db145bd8d5a41e189d/core/src/saros/net/internal/XMPPTransmitter.java#L131

and thought that synchronized doesn't look good.

I was thinking that I don't want to hang the ui on changing for example the skype name in prefs and remembered you mentioning this https://github.com/saros-project/saros/pull/630#discussion_r312391180

Could you explain where do you see something leaking, I don't see what you mean by internal stuff.

srossbach replied:
```sendPacket``` do not block even it is synchronized. The only reason the DispatchThread is visible outside the network package is so skip serialization if you send messages to yourself (the only access in the ActitvitySequencer). The only thing that had to be synchronized is the setting and sending. Otherwise you could trigger a send on contact update, generate a packet which is not send yet, then you set something and this packet is transmitted and afterwards the old packet is transmitted (guess that will never happen but theoretically it could).

stefaus replied:
Ok, I removed the DispatchThread.

srossbach commented:
> 
> 
> > Changing the version comparison logic should not be part of this PR.
> 
> I could but it shouldn't be merged without theses other changes, because this whole pr breaks compatibility multiple times (remove of COMPATIBILITY_KEY, rename message, remove of servicediscovery, xstream maps).

I do not think it matters what patch will break the versioning. Looking at the first 2 commits (that should be part of its own PR) of this PR I think it is sufficient to just set the COMPATIBILITY_KEY value in the response always to: TOO_OLD.

The XML is currently pretty bad formatted and some packet extensions could need an overhaul anyway. So I think it is pretty good to break any compatibility for now. Or did you plan to make a release in the near future (@tobous any plans regarding IntelliJ ?)

saros-infrastructure commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 4
- Added 2
           

Complexity increasing per file
==============================
- core/src/saros/communication/extensions/InfoExchangeExtension.java  2
- core/src/saros/versioning/ClientInfo.java  4
         

Complexity decreasing per file
==============================
+ eclipse/src/saros/ui/eventhandler/ServerPreferenceHandler.java  -1
+ eclipse/src/saros/communication/SkypeManager.java  -2
+ server/src/saros/server/net/ServerFeatureAdvertiser.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5072803&bid=16689379)

tobous commented:
@stefaus For future reference: When you create a stacked pull request (i.e. a PR not based on the master), you have to tell codacy to analyze the base branch (see [settings](https://app.codacy.com/manual/Saros/saros/settings/branches)) beforehand as it otherwise can't run for the PR.

If you forgot, you can simply restart the codacy run after enabling the base branch analysis.

I have already done this for this PR.

srossbach commented on core/src/saros/versioning/VersionManager.java:
```diff
@@ -27,188 +32,144 @@
-                        && exchangeID.equals(versionExchange.get(ID_KEY));
-                  }
-                }));
+  private void sendVersionInfo(XMPPContact contact) {
```
Small info that non Saros contacts are ignored would be nice.

stefaus replied:
done

srossbach commented on core/src/saros/versioning/VersionManager.java:
```diff
@@ -27,188 +32,144 @@
   public VersionManager(
-      @SarosVersion String version, final IReceiver receiver, final ITransmitter transmitter) {
-
+      @SarosVersion String version,
```
As @m273d15 already said, you are mixing concerns. This no longer just a VersionManager. In fact the whole versionlogic is now more or less a one-liner in that class.

stefaus replied:
As I have written: >Not included now for readability: would rename VersionManager to InfoManager and the package saros.versioning to saros.communication.info

I added a Commit renaming the Class and package.

srossbach commented on eclipse/src/saros/communication/SkypeManager.java:
```diff
@@ -4,199 +4,48 @@
     this.preferenceStore = preferenceStore;
-    this.connectionService.addListener(this);
+
+    versionManager.setLocalInfo(PreferenceConstants.SKYPE_USERNAME, getLocalSkypeName());
```
Seems we have to implement startable to the container. Actually this is "heavy" work, at least it seems so, because nothing it actually send at that point anyway. Maybe add an TODO ?

stefaus replied:
Do you mean for the `preferenceStore.addPropertyChangeListener`?
We could add it but does it have a impact?

saros-infrastructure commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 3
           

Complexity increasing per file
==============================
- core/src/saros/communication/info/ClientInfo.java  4
- core/src/saros/communication/info/VersionCompatibilityResult.java  1
- core/test/junit/saros/communication/info/InfoManagerTest.java  1
- core/src/saros/communication/info/Compatibility.java  4
- eclipse/src/saros/ui/eventhandler/ServerPreferenceHandler.java  2
- core/src/saros/communication/info/Version.java  11
- core/src/saros/communication/extensions/InfoExchangeExtension.java  2
- core/src/saros/communication/info/InfoManager.java  5
         

Complexity decreasing per file
==============================
+ server/src/saros/server/net/ServerFeatureAdvertiser.java  -1
+ eclipse/src/saros/communication/SkypeManager.java  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5072803&bid=16689379)

stefaus commented on core/src/saros/net/xmpp/contact/DiscoveryService.java:
```diff
@@ -1,114 +0,0 @@
- * <p>All methods are non-blocking and run in a separate thread, which should be stopped after usage
- * via {@link #stop()}.
- */
-class DiscoveryService {
```
The old DiscoveryManager is already removed since #806

saros-infrastructure commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 3
           

Complexity increasing per file
==============================
- core/src/saros/communication/info/Version.java  11
- core/src/saros/communication/info/VersionCompatibilityResult.java  1
- core/src/saros/communication/info/Compatibility.java  4
- core/test/junit/saros/communication/info/InfoManagerTest.java  1
- eclipse/src/saros/ui/eventhandler/ServerPreferenceHandler.java  2
- core/src/saros/communication/extensions/InfoExchangeExtension.java  2
- core/src/saros/communication/info/InfoManager.java  5
- core/src/saros/communication/info/ClientInfo.java  4
         

Complexity decreasing per file
==============================
+ eclipse/src/saros/communication/SkypeManager.java  -2
+ server/src/saros/server/net/ServerFeatureAdvertiser.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5072803&bid=16689379)

saros-infrastructure commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 3
           

Complexity increasing per file
==============================
- eclipse/src/saros/ui/eventhandler/ServerPreferenceHandler.java  2
- core/src/saros/communication/info/Version.java  11
- core/src/saros/communication/extensions/InfoExchangeExtension.java  2
- core/src/saros/communication/info/InfoManager.java  5
- core/src/saros/communication/info/Compatibility.java  4
- core/src/saros/communication/info/VersionCompatibilityResult.java  1
- core/test/junit/saros/communication/info/InfoManagerTest.java  1
- core/src/saros/communication/info/ClientInfo.java  4
         

Complexity decreasing per file
==============================
+ eclipse/src/saros/communication/SkypeManager.java  -2
+ server/src/saros/server/net/ServerFeatureAdvertiser.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5072803&bid=16689379)

srossbach commented on core/src/saros/communication/info/InfoManager.java:
```diff
@@ -0,0 +1,182 @@
+ *
+ * <p>This class compares if local and remote version (not checking qualifier) are the same.
+ *
+ * <p>If you want to implement backward compatibility in a later version, as a suggestion:
```
This is not very helpful as Javadoc. Guess it is better to just use a simple comment here.

stefaus replied:
done

srossbach commented on core/src/saros/communication/info/InfoManager.java:
```diff
@@ -0,0 +1,182 @@
+  public VersionCompatibilityResult determineVersionCompatibility(JID rqJID) {
+    Version remoteVersion =
+        contactsService
+            .getContact(rqJID.getRAW())
```
It doesn't seem to matter if you use a FQ JID or not. I guess you could also use the FQJID of an instant messager and would get the Saros version as long as the user is also connected with a Saros client.

stefaus replied:
yes, changed to jid

saros-infrastructure commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 3
           

Complexity increasing per file
==============================
- core/src/saros/communication/info/ClientInfo.java  4
- core/src/saros/communication/info/VersionCompatibilityResult.java  1
- core/src/saros/communication/info/Compatibility.java  4
- eclipse/src/saros/ui/eventhandler/ServerPreferenceHandler.java  2
- core/src/saros/communication/info/InfoManager.java  5
- core/src/saros/communication/info/Version.java  11
- core/src/saros/communication/extensions/InfoExchangeExtension.java  2
- core/test/junit/saros/communication/info/InfoManagerTest.java  1
         

Complexity decreasing per file
==============================
+ server/src/saros/server/net/ServerFeatureAdvertiser.java  -1
+ eclipse/src/saros/communication/SkypeManager.java  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5072803&bid=16689379)

saros-infrastructure commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 3
           

Complexity increasing per file
==============================
- core/src/saros/communication/info/Version.java  11
- eclipse/src/saros/ui/eventhandler/ServerPreferenceHandler.java  2
- core/src/saros/communication/info/ClientInfo.java  4
- core/src/saros/communication/info/Compatibility.java  4
- core/test/junit/saros/communication/info/InfoManagerTest.java  1
- core/src/saros/communication/info/VersionCompatibilityResult.java  1
- core/src/saros/communication/info/InfoManager.java  5
- core/src/saros/communication/extensions/InfoExchangeExtension.java  2
         

Complexity decreasing per file
==============================
+ eclipse/src/saros/communication/SkypeManager.java  -2
+ server/src/saros/server/net/ServerFeatureAdvertiser.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5072803&bid=16689379)

tobous commented on server/src/saros/server/net/ServerFeatureAdvertiser.java:
```diff
@@ -1,46 +1,19 @@
-
-    discoveryManager.addFeature(ResourceFeature.SAROS_SERVER.getIdentifier());
+  public ServerFeatureAdvertiser(VersionManager versionManager) {
+    versionManager.setLocalInfo(PreferenceConstants.SERVER_SUPPORT, Boolean.TRUE.toString());
```
Don't you already do this in the `ServerPreferenceHandler`? Why is this logic present twice?

Or is the `ServerPreferenceHandler` the unnecessary one? According to my IDE, it is never referenced/used.

tobous commented on core/src/saros/net/xmpp/XMPPConnectionService.java:
```diff
@@ -272,6 +284,15 @@ public synchronized void connect(
 
+  private void sendAvailablePresenceWithClientIdentifier() {
+    String ver = EntityCapsManager.getInstanceFor(connection).getCapsVersion();
+    CapsExtension caps = new CapsExtension(XMPP_CLIENT_IDENTIFIER, ver, "sha-1");
```
Move hashing algorithm choice to static constant to make it easier to notice when looking at the class?

stefaus replied:
done

tobous commented on core/src/saros/net/xmpp/contact/XMPPContact.java:
```diff
@@ -226,19 +205,15 @@ boolean setNickname(String newNickname) {
-            fullJid,
-            status,
-            oldResource.map(r -> r.features).orElse(EnumSet.noneOf(ResourceFeature.class))));
+    resources.stream().filter(r -> compareRawJid(r.fullJid, fullJid)).map(resources::remove);
```
Does removing an element from a collection inside a stream operation work or can this also lead to `ConcurrentModificationExceptions`?

Use `removeIf()` instead?

Also, the new logic removes all matching entries while the old logic only removed the first found entry. Does this make a difference or is there always only one match?

stefaus replied:
There is always only one match, it worked but was a quick change. Changed to removeIf

tobous commented on core/test/junit/saros/communication/info/InfoManagerTest.java:
```diff
@@ -19,28 +19,29 @@
-
-  private VersionManager versionManagerRemote;
-  private VersionManager versionManagerLocal;
+  private XMPPContactsService aliceContactsService;
```
Make local variable in `init` method as it is not used otherwise?

saros-infrastructure commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 2
           

Complexity increasing per file
==============================
- core/src/saros/communication/extensions/InfoExchangeExtension.java  2
- eclipse/src/saros/ui/eventhandler/ServerPreferenceHandler.java  2
- core/test/junit/saros/communication/info/InfoManagerTest.java  1
- core/src/saros/communication/info/Compatibility.java  4
- core/src/saros/communication/info/InfoManager.java  5
- core/src/saros/communication/info/ClientInfo.java  4
- core/src/saros/communication/info/Version.java  11
- core/src/saros/communication/info/VersionCompatibilityResult.java  1
         

Complexity decreasing per file
==============================
+ server/src/saros/server/net/ServerFeatureAdvertiser.java  -1
+ eclipse/src/saros/communication/SkypeManager.java  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5072803&bid=16689379)

saros-infrastructure commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 1
           

Complexity increasing per file
==============================
- core/src/saros/communication/info/InfoManager.java  5
- core/src/saros/communication/info/VersionCompatibilityResult.java  1
- core/src/saros/communication/extensions/InfoExchangeExtension.java  2
- core/src/saros/communication/info/ClientInfo.java  4
- core/src/saros/communication/info/Compatibility.java  4
- core/test/junit/saros/communication/info/InfoManagerTest.java  1
- core/src/saros/communication/info/Version.java  11
         

Complexity decreasing per file
==============================
+ server/src/saros/server/net/ServerFeatureAdvertiser.java  -1
+ eclipse/src/saros/communication/SkypeManager.java  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5072803&bid=16689379)

stefaus commented on core/src/saros/communication/extensions/VersionExchangeExtension.java:
```diff
@@ -19,30 +20,19 @@
-  public String get(String key) {
-    return data.get(key);
+  public Map<String, String> getData() {
+    return data;
```
changed

stefaus commented on core/src/saros/versioning/VersionManager.java:
```diff
@@ -1,214 +1,145 @@
+        handleInfo(packet);
+      };
+
+  private final IContactsUpdate contactsUpdateListener =
```
yes, it will receive an update for the old resource going offline, and the new becoming available.

stefaus commented on core/src/saros/versioning/VersionManager.java:
```diff
@@ -102,6 +105,38 @@ public VersionCompatibilityResult determineVersionCompatibility(JID jid) {
+
+    if (remoteInfo.isEmpty()) return;
+    // send new info to all known online contacts
+    Enumeration<XMPPContact> contacts = remoteInfo.keys();
```
no, changed it

stefaus commented on eclipse/src/saros/communication/SkypeManager.java:
```diff
@@ -4,199 +4,48 @@
     this.preferenceStore = preferenceStore;
-    this.connectionService.addListener(this);
+
+    versionManager.setLocalInfo(PreferenceConstants.SKYPE_USERNAME, getLocalSkypeName());
```
I added a check, but don't think its important here.

Same doesn't work for the listener because we want to update if the skypename was removed.

m273d15 commented on core/src/saros/communication/info/InfoManager.java:
```diff
@@ -0,0 +1,179 @@
+ * <p>This class compares if local and remote version (not checking qualifier) are the same.
+ */
+@Component(module = "core")
+public class InfoManager {
```
What do you think about splitting the class into two classes:
* InfoManager providing the general methods for exchanging information
* VersionManager using the InfoManager to provide methods for exchanging the versions and determining compatibility

srossbach replied:
I would suggest the same as.

stefaus replied:
Yes, that's a good idea, and I've changed it accordingly.

saros-infrastructure commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 3
- Added 8
           

Complexity increasing per file
==============================
- core/src/saros/communication/extensions/InfoExchangeExtension.java  2
- core/src/saros/communication/InfoManager.java  5
         

Complexity decreasing per file
==============================
+ server/src/saros/server/net/ServerFeatureAdvertiser.java  -1
+ core/src/saros/versioning/VersionManager.java  -2
+ eclipse/src/saros/communication/SkypeManager.java  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5072803&bid=16689379)

saros-infrastructure commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 3
- Added 1
           

Complexity increasing per file
==============================
- core/src/saros/communication/extensions/InfoExchangeExtension.java  2
- core/src/saros/communication/InfoManager.java  5
         

Complexity decreasing per file
==============================
+ server/src/saros/server/net/ServerFeatureAdvertiser.java  -1
+ eclipse/src/saros/communication/SkypeManager.java  -2
+ core/src/saros/versioning/VersionManager.java  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5072803&bid=16689379)

stefaus commented on core/src/saros/communication/InfoManager.java:
```diff
@@ -0,0 +1,140 @@
+ */
+public class InfoManager {
+  private class ClientInfo {
+    private final Map<String, String> infos;
```
I removed in the mean time because it has not a big value and I use ClientInfo mainly to avoid `Map<XMPPContact, Map<String, String>>` but yes probably better to keep it.

stefaus commented on core/src/saros/communication/InfoManager.java:
```diff
@@ -0,0 +1,140 @@
+ * String)}.
+ */
+public class InfoManager {
+  private class ClientInfo {
```
done

stefaus commented on core/src/saros/communication/InfoManager.java:
```diff
@@ -0,0 +1,140 @@
+    Optional<XMPPContact> contact = contactsService.getContact(packet.getFrom());
+    if (!contact.isPresent()) return;
+
+    InfoExchangeExtension versionInfo = InfoExchangeExtension.PROVIDER.getPayload(packet);
```
done

srossbach commented on core/src/saros/communication/InfoManager.java:
```diff
@@ -0,0 +1,144 @@
+    try {
+      transmitter.sendPacket(packet);
+    } catch (IOException e) {
+      log.error("could not send version response to " + sarosJid, e);
```
error message is wrong

stefaus replied:
done

saros-infrastructure commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 3
- Added 1
           

Complexity increasing per file
==============================
- core/src/saros/communication/extensions/InfoExchangeExtension.java  2
- core/src/saros/communication/InfoManager.java  5
         

Complexity decreasing per file
==============================
+ core/src/saros/versioning/VersionManager.java  -2
+ eclipse/src/saros/communication/SkypeManager.java  -2
+ server/src/saros/server/net/ServerFeatureAdvertiser.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5072803&bid=16689379)

saros-infrastructure commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 3
- Added 1
           

Complexity increasing per file
==============================
- core/src/saros/communication/extensions/InfoExchangeExtension.java  2
- core/src/saros/communication/InfoManager.java  5
         

Complexity decreasing per file
==============================
+ core/src/saros/versioning/VersionManager.java  -2
+ eclipse/src/saros/communication/SkypeManager.java  -2
+ server/src/saros/server/net/ServerFeatureAdvertiser.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5072803&bid=16689379)

saros-infrastructure commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 3
- Added 1
           

Complexity increasing per file
==============================
- core/src/saros/communication/InfoManager.java  5
- core/src/saros/communication/extensions/InfoExchangeExtension.java  2
         

Complexity decreasing per file
==============================
+ eclipse/src/saros/communication/SkypeManager.java  -2
+ core/src/saros/versioning/VersionManager.java  -2
+ server/src/saros/server/net/ServerFeatureAdvertiser.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5072803&bid=16689379)
